### PR TITLE
[Sema] Synthesize default values for memberwise init

### PR DESCRIPTION
**Pull request by Azoy**
Previously we were synthesizing something like the following:
```swift
struct Dog {
  var age = 0
  var name = ""

  // compiler synthesized
  init(age: Int, name: String)

  // compiler synthesized
  init() {}
}
```

This patch is a QoL change to the way we synthesize the memberwise constructor for structures.

```swift
struct Dog {
  var age = 0
  var name = ""

  // compiler synthesized
  init(age: Int = 0, name: String = "")
}
```

Swift Evolution review: https://forums.swift.org/t/se-0242-synthesize-default-values-for-the-memberwise-initializer/20618

cc: @jrose-apple @harlanhaskins

### CODE REVIEW

**Comment by slavapestov** [comment]
@Azoy Instead of moving the initializer into a helper function, a different strategy would be to add a new DefaultArgumentKind that simply points at the VarDecl.

VarDecls with an initial value already emit a generator function in SILGen, and this function can itself be used as the default argument for the constructor.

This way you avoid emitting a helper function entirely.

File: lib/AST/Expr.cpp
Comment (harlanhaskins) [inline lib/AST/Expr.cpp]: Should this operate on `this` instead of a parameter?

Comment (Azoy) [reply lib/AST/Expr.cpp]: I initially did it that way, but in ParsePattern I had to write something like:
```cpp
if (param.DefaultArg) {
  kind = param.DefaultArg->getDefaultArgKind();
} else {
  kind = DefaultArgumentKind::None;
}
```
This isn't a huge deal, but it certainly seemed nicer to just write `Expr::getDefaultArgKind(param.DefaultArg);` instead.

Comment (harlanhaskins) [reply lib/AST/Expr.cpp]: I personally prefer the former, it makes it slightly more clear what you're doing. Also you could just write

```c++
auto kind = param.DefaultArg ?
            param.DefaultArg->getDefaultArgKind :: DefaultArgKind::None;
```

**Comment by Azoy** [comment]
@slavapestov what value can I give the parameter for 1. the type checker to be happy and 2. be able to call the variable initializer?

**Comment by slavapestov** [comment]
@Azoy Sorry, I'm not sure what you mean by "parameter"

**Comment by Azoy** [comment]
@slavapestov Apologies if something here isn't correct, I'm new to the SILGen codebase, but I think for the most part this looks ok?
```swift
struct X {
  var y = 0
}
```
The memberwise constructor's default argument looks like this:
```sil
// default argument 0 of X.init(y:)
sil hidden @$s4test1XV1yACSi_tcfcfA_ : @$convention(thin) () -> Int {
bb0:
  // function_ref variable initialization expression of X.y
  %0 = function_ref @$s4test1XV1ySivpfi : @$convention(thin) () -> Int // user: %1
  %1 = apply %0() : @$convention(thin) () -> Int //user: %2
  return %1 : $Int
} // end sil function '$s4test1XV1yACSi_tcfcfA_'
```
Currently the stored property kind is printed as the variable name, so @jrose-apple or @harlanhaskins might have some input on how to properly print it.

**Comment by slavapestov** [comment]
I’ll take a look tomorrow. I don’t think it matters how we print these default arguments since memberwise initializers are never public.

Also your PR needs tests :)

**Comment by Azoy** [comment]
@slavapestov I've run into some interesting behavior with optionals concerning with the creation of the implicit constructor(s).
Swift 4.2:
```swift
struct X {
  var y: Int?
  // Compiler synthesized
  init(y: Int?)
  // Compiler synthesized
  init() { return }
}

let x = X()
```
Would it make sense to assign the default value of `y: Int?` in the memberwise constructor to `nil`?
```swift
struct X {
  var y: Int?
  // Compiler synthesized
  init(y: Int? = nil)
}

let x = X()
```

**Comment by slavapestov** [comment]
@Azoy, yes, properties with optional type have a "default" initial value of nil. I suggested to swift-evolution once to remove this edge case, but there was push back so we're stuck with it. :)

**Comment by Azoy** [comment]
I'm having some trouble with some of the SIL being generated for the default argument for some types.
Given:
```swift
import Foundation
struct X {
  var y = Data(bytes: [])
}
```
Default argument function:
```sil
SIL verification failed: entry point has wrong number of arguments: entry->args_size() == (fnConv.getNumIndirectSILResults() + fnConv.getNumParameters())
In function:
// default argument 0 of X.init(y:)
sil hidden @$s6REPL_11XV1yAC10Foundation4DataV_tcfcfA_ : $@convention(thin) () -> @out Data {
bb0:
  // function_ref variable initialization expression of X.y
  %0 = function_ref @$s6REPL_11XV1y10Foundation4DataVvpfi : $@convention(thin) () -> @out Data // user: %2
  %1 = alloc_stack $Data                          // users: %3, %2
  %2 = apply %0(%1) : $@convention(thin) () -> @out Data
  return %1 : $*Data                              // id: %3
} // end sil function '$s6REPL_11XV1yAC10Foundation4DataV_tcfcfA_'
```
I see what it's trying to do with ResultPlan noticing the indirect `@out Data` and wanting to create a temporary and return that, but obviously its tripping over trying to apply the temporary to the variable initializer with 0 arguments. Am I setting up the emitApplyOfStoredPropertyInitializer incorrectly?

File: lib/Sema/TypeCheckDecl.cpp
Comment (Azoy) [reply lib/Sema/TypeCheckDecl.cpp]: I hadn't even thought of efficiency when deciding this. I agree that default is more efficient by far.

File: lib/Sema/CodeSynthesis.cpp
Comment (Azoy) [reply lib/Sema/CodeSynthesis.cpp]: I don't think we can because given:
```swift
struct X {
  var y: Int?
  var z = 0
}

let x = X(z: 16)
```
Type checker catches this before this is handed off to SILGen because memberwise initializer produced is:
```swift
init(y: Int?, z: Int = z)
```
Which produces the error: `cannot invoke initializer for type 'X' with an argument list of type '(z: Bool)'`.

**Comment by Azoy** [comment]
@harlanhaskins Could you take a look at the latest commit? I haven't run validation tests yet, but the only failing test I'm getting is with `SILOptimizer/exclusivity_static_diagnostics_inlined.swift` error: `cannot use instance member 'someV' as a default parameter`. I'm not entirely sure how I'm getting that diagnostic, but I think its because we're printing the instance member name as a the default text.

**Comment by Azoy** [comment]
@slavapestov is there a specific test file for the memberwise constructor that I'm missing, or is it some other test file?

**Comment by natecook1000** [comment]
@Azoy This looks great, glad to see it coming up for review! Would you mind resolving the conflict? I'd like to build a toolchain with this feature.

**Comment by Azoy** [comment]
@natecook1000 done and done!

**Comment by natecook1000** [comment]
@swift-ci Please build toolchain

**Comment by natecook1000** [comment]
@swift-ci Please build toolchain macOS platform

**Comment by natecook1000** [comment]
@swift-ci Please build toolchain macOS platform

**Comment by natecook1000** [comment]
@shahmishal Trying to build a macOS toolchain — am I doing something wrong here?

**Comment by shahmishal** [comment]
Github is having issues, which is causing CI to fail. 

https://www.githubstatus.com

screen shot 2019-02-19 at 10 18 13 am

**Comment by shahmishal** [comment]
@swift-ci Please build toolchain macOS platform

**Comment by shahmishal** [comment]
@swift-ci Please build toolchain macOS platform

**Comment by brentdax** [comment]
@swift-ci please test source compatibility

**Comment by brentdax** [comment]
@swift-ci please smoke test compiler performance

**Comment by brentdax** [comment]
I'm running these additional tests because we accepted several proposals in the Swift 5 cycle which seemed like they should have no source compatibility impact, but in practice they expanded overload sets in ways that were very challenging for the type checker. Let's go in with our eyes open.

**Comment by brentdax** [comment]
> Unexpected test results, excluded stats for ReactiveCocoa

ReactiveCocoa appears to have failed to build with both the old and new compilers, so this isn't a concern.

> | name | old | new | delta | delta_pct |
> | ---- | --- | --- | ----- | --------- |
> | Sema.NumConstraintScopes | 38,798 | 38,840 | 42 | 0.11% |

This shows a small (well below threshold) effect on the type checker, but IIRC, an increase in the number of constraint scopes is not necessarily a bad thing. Right, @xedin?

> | name | old | new | delta | delta_pct |
> | ---- | --- | --- | ----- | --------- |
> | Frontend.NumInstructionsExecuted | 141,114,637,538 | 141,135,478,890 | 841,352 | 0.01% |
> | time.swift-driver.wall | 26.3s | 26.4s | 118.1ms | 0.45% |

These effects are also below threshold and don't look too severe. Might be a good idea to see if there's any low-hanging performance fruit in this PR, though.

A smoke performance test isn't a full performance test, of course, but I don't see anything showstopping yet.

> | name | old | new | delta | delta_pct |
> | ---- | --- | --- | ----- | --------- |
> | LLVM.NumLLVMBytesOutput | 7,136,792 | 7,136,780 | -12 | -0.0% |

...alrighty, then. A day may come when I understand how this shit happens, but it is not this day.

**Comment by brentdax** [comment]
Source compatibility looks good too.

> Passed: 261
> Failed: 0
> XFailed: 28
> UPassed: 0
> Total: 289

**Comment by Azoy** [comment]
@slavapestov could it make sense to move the logic in emitDefaultArgument to emitApplyOfDefaultArgGenerator (whether to emit apply to default arg generator or stored property initializer)? I ask because I keep looking at the logic there and keep wondering if there is a cleaner way of doing it.

**Comment by tkremenek** [comment]
@swift-ci clean test

File: lib/SILGen/SILGenApply.cpp
Comment (slavapestov) [inline lib/SILGen/SILGenApply.cpp]: You should be able to say subs = F.getForwardingSubstitutions()

File: lib/Sema/CodeSynthesis.cpp
Comment (slavapestov) [reply lib/Sema/CodeSynthesis.cpp]: I'm still not sure you need to actually build the NilLiteralExpr here and set the default argument kind to NilLiteral. Since the argument is emitted in SILGen as a call to the property initializer, you should still be able to use DefaultArgumentKind::StoredProperty here and avoid building the NilLiteralExpr.

File: lib/Sema/CodeSynthesis.cpp
Comment (slavapestov) [inline lib/Sema/CodeSynthesis.cpp]: This method is getting a bit long -- do you mind splitting out the default argument logic?

File: include/swift/Serialization/ModuleFormat.h
Comment (slavapestov) [inline include/swift/Serialization/ModuleFormat.h]: Since you changed the module format please also increment the minor version number at the top of this file

**Review by slavapestov — COMMENTED** [review]
This is almost ready to go!

**Comment by slavapestov** [comment]
@Azoy This also needs SILGen tests. Maybe start by extending test/SILGen/default_constructor.swift, or creating a new test file. Make sure to test non-generic and generic types.

**Comment by Azoy** [comment]
@slavapestov We need Sema tests as well no? Do you have any idea where the best place to add those tests would be? I've gone searching up and down looking for anything memberwise specific, but couldn't find anything. Maybe just make a new test file?

File: lib/SILGen/SILGenApply.cpp
Comment (Azoy) [reply lib/SILGen/SILGenApply.cpp]: Hmm, this might be a bit more complicated than I thought. A simple example like:
```swift
struct X<T> {
  var y: T
  var z = false
}

let x = X(y: 3)
```
The variable initializer that we call in this case is `z`'s whose function signature is:
```sil
$@convention(thin) <T> () -> Bool
```
We can't call this in main because the application uses a generic that isn't a part of main's signature. i.e.
```sil
// function_ref variable initialization expression of X.z
%11 = function_ref @$s4main1XV1zSbvpfi : $@convention(thin) <τ_0_0> () -> Bool // user: %12
%12 = apply %11<T>() : $@convention(thin) <τ_0_0> () -> Bool // user: %14
```
Which crashes during verification because its looking for the `T` generic in main's signature, which there is not. Any ideas?

Comment (Azoy) [reply lib/SILGen/SILGenApply.cpp]: Also, does it make sense that the variable initializer in this case has a generic in its signature? Could this simply be `$@convention(thin) () -> Bool`?

**Comment by Azoy** [comment]
@slavapestov I would love it if you could take another look and suggest some more test cases if you have any. I believe we're close!

File: lib/Sema/CodeSynthesis.cpp
Comment (Azoy) [inline lib/Sema/CodeSynthesis.cpp]: To shed some light on this change, the way printing `StoredProperty` works is by getting the parent initializer, but for `T?` cases there is no initializer. We don't build out any `NilLiteralExpr`, but we set the arg kind so that printing will just give us `nil`.

File: lib/SILGen/SILGenApply.cpp
Comment (slavapestov) [inline lib/SILGen/SILGenApply.cpp]: I think you can remove the preceding 'if'

**Review by slavapestov — APPROVED** [review]

**Comment by slavapestov** [comment]
I think this looks good enough to merge, except for two small things:
- do you mind taking a look at my final comment above?
- Also, do you want to add an entry to the changelog?

**Comment by Azoy** [comment]
@slavapestov is it ok to wait on changelog until it has been merged to the, assuming, 5.1 branch?

**Comment by slavapestov** [comment]
No need to merge to 5.1 branch, we're re-branching soon. You can add the changelog entry in a subsequent PR though, but please try to find time to do it soon if you can!

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by Azoy** [comment]
@DougGregor @slavapestov There’s still a slight issue when it comes to lazy variables that I haven’t figured out yet. We stored the lazy property as the default value, but when we emit the default argument, we want to call the storage’s initialization rather than the lazy property itself (since it has no initializer). From what I’ve found, there’s really not a great way of just asking for the storage and using it, I’ll have to somehow plumb it down, or just iterate the members until I find it. Maybe there’s an easier way of just grabbing the storage variable?

**Comment by DougGregor** [comment]
Note: `test/decl/var/lazy_properties_batch_mode.swift` is failing with an assertion.

**Comment by DougGregor** [comment]
@Azoy ah, there is currently no good way to get the storage variable for a `lazy` property. I just tripped over a TODO that suggests that we add a mapping from the `lazy` property back to the storage variable so it's easy to retrieve.

**Comment by DougGregor** [comment]
Ah, this is the place I found the TODO: https://github.com/apple/swift/blob/master/lib/Sema/CodeSynthesis.cpp#L1424-L1428

**Comment by Azoy** [comment]
@DougGregor @slavapestov I've committed a fix for lazy vars, so tests should pass now. Please take a look at the latest commit and let me know if there is anything else I should adjust or change.

**Comment by tkremenek** [comment]
@swift-ci test

File: lib/Sema/CodeSynthesis.cpp
Comment (slavapestov) [inline lib/Sema/CodeSynthesis.cpp]: I'm worried that this adds some non-determinism to the behavior here. There's no guarantee that the lazy property has been completed at the time you're synthesizing the default argument for `var`.

However, you really don't care one way or another: whether or not the lazy property's storage has been synthesized should not affect the behavior here. In fact, the type of a lazy property parameter to a memberwise init is an Optional of the lazy property's type, therefore the default argument value for a lazy property is the default argument value for the lazy property's _storage_, which is always nil. So you can remove the areLazyAccessorsSynthesized and instead treat the default argument as a nil initialized value here.

It might be tricky to come up with a test case that triggers the incorrect behavior with the code as-written, but one way you might be able to do it is by fiddling with the order of declarations in your Sema test. Try calling the memberwise init before the struct declaration in the file, and it might be enough to trigger this code path before the lazy variable is completed (which happens late).

File: lib/Sema/CodeSynthesis.cpp
Comment (slavapestov) [reply lib/Sema/CodeSynthesis.cpp]: ... in fact this already appears to correctly handle the lazy property case. Maybe just remove the `areLazyAccessorsSynthesized` check above and see what happens?

Also, can you restructure this code a bit so that setDefaultArgumentKind() is only called once? It's a bit confusing to see it overwritten later. I'm guessing you don't really need to call setStoredProperty() either in the NilLiteral case?

File: lib/SILGen/SILGenApply.cpp
Comment (slavapestov) [inline lib/SILGen/SILGenApply.cpp]: Can this happen? In maybeAddMemberwiseDefaultArg() it looks like you set the default argument kind of lazy properties to NilLiteral.

My naive understanding of SILGenApply suggests that in TupleShuffleArgEmitter::emitDefaultArgsAndFinalize(), this causes Sema to create a `TupleShuffleExpr::CallerDefaultInitialize` at the call site, so we won't end up here at all. However I might be wrong about that.

File: lib/Sema/CodeSynthesis.cpp
Comment (slavapestov) [inline lib/Sema/CodeSynthesis.cpp]: Can you put the Lazy attribute check outside of `isa<OptionalType>`?

**Review by slavapestov — COMMENTED** [review]
Sorry to come back with more review feedback at this stage; I also missed the 'lazy' case and I want to make sure we get it right because 'lazy' has tripped all of us up before.

File: lib/Sema/CodeSynthesis.cpp
Comment (Azoy) [reply lib/Sema/CodeSynthesis.cpp]: The reason for this check is because lazy variables whose accessors have been synthesized have removed the `hasStorage` check from the variable as well which is an assertion in `isDefaultInitializable`. Those that have not been completed still pass the following check because they are still considered to have their storage.

Comment (slavapestov) [reply lib/Sema/CodeSynthesis.cpp]: I think its less error-prone to handle both cases uniformly by only checking if `hasStorage()` on properties that are not lazy. If the property has a LazyAttr, you proceed regardless.

Comment (Azoy) [reply lib/Sema/CodeSynthesis.cpp]: We don't really want to just check `hasStorage` either because variables without initial expressions proceed and we want to return early if that's not the case. Something like:
```cpp
if (!var->getAttrs().hasAttribute<LazyAttr>() &&
    !var->getParentPatternBinding()->isDefaultInitializable())
  return;
```
could work maybe?

Comment (slavapestov) [reply lib/Sema/CodeSynthesis.cpp]: Yes, that looks fine.

**Comment by Azoy** [comment]
@slavapestov ok, I think this is the one! I made it so that we check for `NilLiteral` cases before we set stored property and just return if it's the case.

File: lib/SILGen/SILGenApply.cpp
Comment (slavapestov) [inline lib/SILGen/SILGenApply.cpp]: Does this work across modules? Normally you can’t call a memberwise initializer since they’re internal, but if a module is built with -enable-testing and imported with @testable import you can. I might be wrong but I don’t see ParamDecl::getStoredProperty() serialized anywhere.

File: lib/SILGen/SILGenApply.cpp
Comment (slavapestov) [inline lib/SILGen/SILGenApply.cpp]: Is this still needed?

File: lib/SILGen/SILGenApply.cpp
Comment (slavapestov) [inline lib/SILGen/SILGenApply.cpp]: This feels a bit fragile. Is there a way to get the ParamDecl from the DefaultArgumentStorage instead?

**Review by slavapestov — APPROVED** [review]
The new version looks great! I still have a small cleanup suggestion, and one concern related to serialization with -enable-testing but I think we should merge this now, and address the below in a follow-up PR.

**Comment by slavapestov** [comment]
@Azoy do you mind updating the PR description now that the proposal has been accepted and the implementation has moved away from synthesizing  closure?

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
Looks like CI is having problems. I'm on vacation until Tuesday; @DougGregor or @tkremenek will help you get this merged in the mean time.

File: lib/SILGen/SILGenApply.cpp
Comment (Azoy) [reply lib/SILGen/SILGenApply.cpp]: I tried to spin up a quick test to test the current behavior, and I don't believe that we currently deserialize any default value from a module, but rather just the arg kind and the string representation. I'd be curious to hear how we could go about setting the stored property arg after we deserialize (maybe just take param name and look for the corresponding property?). Right now this emits an apply to the default arg generator (which there is none, so I believe the linker will complain...).

**Comment by natecook1000** [comment]
@swift-ci Please test Linux platform

File: lib/SILGen/SILGenApply.cpp
Comment (slavapestov) [reply lib/SILGen/SILGenApply.cpp]: I think the right solution is to change the code they serializes a ParamDecl to serialize a cross reference to the corresponding VarDecl. I would try to avoid a name-based lookup here. @jrose-apple might have more suggestions

Comment (jrose-apple) [reply lib/SILGen/SILGenApply.cpp]: I think emitting an apply to the default arg generator is the correct behavior. The fact that the expression comes from a stored property shouldn't matter except when generating the function.

Comment (Azoy) [reply lib/SILGen/SILGenApply.cpp]: Emitting a default arg generator in this case doesn't work because the module with the stored property default arg doesn't emit a default arg generator function. So when we go to deserialize a stored property default arg, if we emit a default arg generator apply then we're going to be linking against undefined symbols.

**Comment by tkremenek** [comment]
@slavapestov has approved, so I am merging.

**Comment by tkremenek** [comment]
@azoy: Please create a PR against `swift-5.1-branch` to pull this into 5.1

File: lib/SILGen/SILGenApply.cpp
Comment (jrose-apple) [reply lib/SILGen/SILGenApply.cpp]: > Emitting a default arg generator in this case doesn't work because the module with the stored property default arg doesn't emit a default arg generator function.

Hm. Why doesn't it? Isn't this a bug? (under `-enable-testing`, anyway)

**Comment by tkremenek** [comment]
I just noticed the recent commentary on this PR.  I think this is OK to merge now for incremental development, but let's address the remaining concerns in a separate PR.  A 5.1 PR can contain changes from both PRs.

**Comment by rintaro** [comment]
When creating PR for `swift-5.1-branch` for this, please include #23564 as well.

**Comment by tkremenek** [comment]
@Azoy: is there a 5.1 PR for this yet?

**Comment by Azoy** [comment]
@tkremenek No, sorry not yet. I'm working with Jordan and Slava to get https://github.com/apple/swift/pull/23648 merged. Once that's merged (which is hopefully tonight), I will submit a 5.1 PR.

**Comment by tkremenek** [comment]
@azoy https://github.com/apple/swift/pull/23648 has been merged

**Pull request merged**
